File: src/main/java/com/refinedmods/refinedstorage/screen/grid/filtering/CraftableGridFilter.java
package com.refinedmods.refinedstorage.screen.grid.filtering;

import com.refinedmods.refinedstorage.screen.grid.stack.IGridStack;

import java.util.function.Predicate;

public class CraftableGridFilter implements Predicate<IGridStack> {
    private final boolean craftable;

    public CraftableGridFilter(boolean craftable) {
        this.craftable = craftable;
    }

    @Override
    public boolean test(IGridStack stack) {
        if (craftable) {
            return stack.isCraftable();
        } else {
            return !stack.isCraftable() && stack.getOtherId() == null;
        }
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/screen/grid/filtering/CraftableGridFilter.kt
package com.refinedmods.refinedstorage.screen.grid.filtering

import com.refinedmods.refinedstorage.screen.grid.stack.IGridStack
import java.util.function.Predicate

class CraftableGridFilter(private val craftable: Boolean) : Predicate<IGridStack> {
    override fun test(stack: IGridStack): Boolean {
        return if (craftable) {
            stack.isCraftable
        } else {
            !stack.isCraftable && stack.otherId == null
        }
    }
}